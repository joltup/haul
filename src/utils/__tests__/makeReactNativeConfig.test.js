/**
 * Copyright 2017-present, Callstack.
 * All rights reserved.
 *
 * @flow
 */

import path from 'path';
import { replacePathsInObject } from 'jest/helpers'; // eslint-disable-line import/no-unresolved
import {
  makeReactNativeConfig,
  injectPolyfillIntoEntry,
} from '../makeReactNativeConfig';

test('creates config from defaults', () => {
  const webpackConfig = require('./fixtures/webpack.config.js');
  const iosConfig = makeReactNativeConfig(
    webpackConfig,
    {
      dev: true,
      root: path.resolve(__dirname, 'fixtures'),
    },
    'ios'
  );

  expect(replacePathsInObject(iosConfig)).toMatchSnapshot('ios config');

  const androidConfig = makeReactNativeConfig(
    webpackConfig,
    {
      dev: true,
      root: path.resolve(__dirname, 'fixtures'),
    },
    'android'
  );

  expect(replacePathsInObject(androidConfig)).toMatchSnapshot('android config');
});

test('merges existing config', () => {
  const webpackConfig = require('./fixtures/webpack.custom.config.js');
  const config = makeReactNativeConfig(
    webpackConfig,
    {
      dev: true,
      root: path.resolve(__dirname, 'fixtures'),
    },
    'ios'
  );

  expect(replacePathsInObject(config)).toMatchSnapshot();
});

describe('injects polyfill into different entries', () => {
  const fakePolyfillPath = 'path/to/polyfill.js';

  test('entry is a string', () => {
    const userEntry = './src/index.js';
    const generatedEntry = injectPolyfillIntoEntry(userEntry, fakePolyfillPath);

    expect(generatedEntry.length).toBe(2);
    expect(generatedEntry[0]).toBe(fakePolyfillPath);
  });

  test('entry is an array', () => {
    const userEntry = ['./src/index.js', './src/module.js'];
    const generatedEntry = injectPolyfillIntoEntry(userEntry, fakePolyfillPath);

    expect(generatedEntry[0]).toBe(fakePolyfillPath);
    expect(generatedEntry.length).toBe(3);
  });

  test('entry is an object', () => {
    const userEntry = {
      entry1: './src/index.js',
      entry2: ['./src/module.js', './src/vendor.js'],
    };
    const expectedEntry1 = [fakePolyfillPath, './src/index.js'];

    const expectedEntry2 = [
      fakePolyfillPath,
      './src/module.js',
      './src/vendor.js',
    ];
    const generatedEntry = injectPolyfillIntoEntry(userEntry, fakePolyfillPath);

    expect(generatedEntry).toMatchObject({
      entry1: expectedEntry1,
      entry2: expectedEntry2,
    });
  });
});
