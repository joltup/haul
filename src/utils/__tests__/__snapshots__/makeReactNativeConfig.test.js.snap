// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates config from defaults: creates config from defaults (configs) 1`] = `
Object {
  "context": "<<REPLACED>>/utils/__tests__/fixtures",
  "entry": Array [
    "<<REPLACED>>/utils/polyfillEnvironment.js",
    "./file.js",
  ],
  "module": Object {
    "rules": Array [
      Object {
        "parser": Object {
          "requireEnsure": false,
        },
      },
      Object {
        "exclude": /node_modules\\\\/\\(\\?!react\\|@expo\\|pretty-format\\|haul\\)/,
        "test": /\\\\\\.js\\$/,
        "use": Array [
          Object {
            "loader": "<<NODE_MODULE>>/thread-loader/dist/cjs.js",
          },
          Object {
            "loader": "<<NODE_MODULE>>/babel-loader/lib/index.js",
            "options": Object {
              "cacheDirectory": true,
              "extends": "<<REPLACED>>/utils/__tests__/fixtures/.babelrc",
              "plugins": Array [
                "<<REPLACED>>/utils/fixRequireIssues.js",
                "<<NODE_MODULE>>/react-hot-loader/babel.js",
                "<<REPLACED>>/hot/babelPlugin.js",
              ],
            },
          },
        ],
      },
      Object {
        "test": /\\\\\\.\\(aac\\|aiff\\|bmp\\|caf\\|gif\\|html\\|jpeg\\|jpg\\|m4a\\|m4v\\|mov\\|mp3\\|mp4\\|mpeg\\|mpg\\|obj\\|otf\\|pdf\\|png\\|psd\\|svg\\|ttf\\|wav\\|webm\\|webp\\)\\$/,
        "use": Object {
          "loader": "<<REPLACED>>/loaders/assetLoader.js",
          "query": Object {
            "bundle": undefined,
            "platform": "ios",
            "root": "<<REPLACED>>/utils/__tests__/fixtures",
          },
        },
      },
    ],
  },
  "name": "ios",
  "output": Object {
    "filename": "index.ios.bundle",
    "path": "<<REPLACED>>/utils/__tests__/fixtures",
    "publicPath": "http://localhost:undefined/",
  },
  "plugins": Array [
    CaseSensitivePathsPlugin {
      "fsOperations": 0,
      "options": Object {},
      "pathCache": Object {},
      "primed": false,
    },
    ProgressPlugin {
      "handler": [Function],
      "profile": undefined,
    },
    DefinePlugin {
      "definitions": Object {
        "__DEV__": true,
        "process.env": Object {
          "NODE_ENV": "\\"development\\"",
        },
      },
    },
    LoaderOptionsPlugin {
      "options": Object {
        "debug": true,
        "minimize": false,
        "test": Object {
          "test": [Function],
        },
      },
    },
    HotModuleReplacementPlugin {
      "fullBuildTimeout": 200,
      "multiStep": undefined,
      "options": Object {},
      "requestTimeout": 10000,
    },
    EvalSourceMapDevToolPlugin {
      "options": Object {
        "module": true,
      },
    },
    NamedModulesPlugin {
      "options": Object {},
    },
  ],
  "resolve": Object {
    "alias": Object {
      "react-proxy": "@zamotany/react-proxy",
    },
    "extensions": Array [
      ".ios.js",
      ".native.js",
      ".js",
    ],
    "mainFields": Array [
      "react-native",
      "browser",
      "main",
    ],
    "plugins": Array [
      HasteResolver {
        "apply": [Function],
      },
      AssetResolver {
        "apply": [Function],
      },
    ],
  },
  "target": "webworker",
}
`;

exports[`creates config from defaults: creates config from defaults (configs) 2`] = `
Object {
  "context": "<<REPLACED>>/utils/__tests__/fixtures",
  "entry": Array [
    "<<REPLACED>>/utils/polyfillEnvironment.js",
    "./file.js",
  ],
  "module": Object {
    "rules": Array [
      Object {
        "parser": Object {
          "requireEnsure": false,
        },
      },
      Object {
        "exclude": /node_modules\\\\/\\(\\?!react\\|@expo\\|pretty-format\\|haul\\)/,
        "test": /\\\\\\.js\\$/,
        "use": Array [
          Object {
            "loader": "<<NODE_MODULE>>/thread-loader/dist/cjs.js",
          },
          Object {
            "loader": "<<NODE_MODULE>>/babel-loader/lib/index.js",
            "options": Object {
              "cacheDirectory": true,
              "extends": "<<REPLACED>>/utils/__tests__/fixtures/.babelrc",
              "plugins": Array [
                "<<REPLACED>>/utils/fixRequireIssues.js",
                "<<NODE_MODULE>>/react-hot-loader/babel.js",
                "<<REPLACED>>/hot/babelPlugin.js",
              ],
            },
          },
        ],
      },
      Object {
        "test": /\\\\\\.\\(aac\\|aiff\\|bmp\\|caf\\|gif\\|html\\|jpeg\\|jpg\\|m4a\\|m4v\\|mov\\|mp3\\|mp4\\|mpeg\\|mpg\\|obj\\|otf\\|pdf\\|png\\|psd\\|svg\\|ttf\\|wav\\|webm\\|webp\\)\\$/,
        "use": Object {
          "loader": "<<REPLACED>>/loaders/assetLoader.js",
          "query": Object {
            "bundle": undefined,
            "platform": "android",
            "root": "<<REPLACED>>/utils/__tests__/fixtures",
          },
        },
      },
    ],
  },
  "name": "android",
  "output": Object {
    "filename": "index.android.bundle",
    "path": "<<REPLACED>>/utils/__tests__/fixtures",
    "publicPath": "http://localhost:undefined/",
  },
  "plugins": Array [
    CaseSensitivePathsPlugin {
      "fsOperations": 0,
      "options": Object {},
      "pathCache": Object {},
      "primed": false,
    },
    ProgressPlugin {
      "handler": [Function],
      "profile": undefined,
    },
    DefinePlugin {
      "definitions": Object {
        "__DEV__": true,
        "process.env": Object {
          "NODE_ENV": "\\"development\\"",
        },
      },
    },
    LoaderOptionsPlugin {
      "options": Object {
        "debug": true,
        "minimize": false,
        "test": Object {
          "test": [Function],
        },
      },
    },
    HotModuleReplacementPlugin {
      "fullBuildTimeout": 200,
      "multiStep": undefined,
      "options": Object {},
      "requestTimeout": 10000,
    },
    EvalSourceMapDevToolPlugin {
      "options": Object {
        "module": true,
      },
    },
    NamedModulesPlugin {
      "options": Object {},
    },
  ],
  "resolve": Object {
    "alias": Object {
      "react-proxy": "@zamotany/react-proxy",
    },
    "extensions": Array [
      ".android.js",
      ".native.js",
      ".js",
    ],
    "mainFields": Array [
      "react-native",
      "browser",
      "main",
    ],
    "plugins": Array [
      HasteResolver {
        "apply": [Function],
      },
      AssetResolver {
        "apply": [Function],
      },
    ],
  },
  "target": "webworker",
}
`;

exports[`merges existing config: merges existing config 1`] = `
Object {
  "context": "<<REPLACED>>/utils/__tests__/fixtures",
  "entry": Array [
    "<<REPLACED>>/utils/polyfillEnvironment.js",
    "/some/file.js",
  ],
  "module": Object {
    "rules": Array [
      Object {
        "parser": Object {
          "requireEnsure": false,
        },
      },
    ],
  },
  "name": "ios",
  "output": Object {
    "filename": "index.ios.bundle",
    "path": "<<REPLACED>>/utils/__tests__/fixtures",
    "publicPath": "http://localhost:undefined/",
  },
  "plugins": Array [],
  "resolve": Object {},
  "target": "webworker",
}
`;
